### ISSUE TITLE
feat: Reduce `put` time

### ISSUE THREAD

**Issue by thomass-dev** [probabl-ai/skore#2140]
- [x] Only push permutations if they are computed ([#2094](https://github.com/probabl-ai/skore/pull/2094))
- [x] Parallelize metric upload ([#2178](https://github.com/probabl-ai/skore/pull/2178))
- [ ] Use the hash generated by `skore` to upload a report ([#2114](https://github.com/probabl-ai/skore/pull/2114))
- [ ] Parallelize media upload ([#2114](https://github.com/probabl-ai/skore/pull/2114))
- [ ] Retrieve media from report's cache generated by `skore`
- [ ] Serialize intelligently reports (#2200)

**Comment by glemaitre** [comment]
@thomass-dev don't hesitate to open sub-issues: I feel that "Serialize intelligently reports" would require some digging and maybe having its own issue before to go to the implementation then it will be easier.

**Comment by auguste-probabl** [comment]
This should be P0, we have some users struggling with this

**Comment by jeromedockes** [comment]
It seems there are a few directions that we have started working on:

1. Deduplicating datasets. Instead of uploading the full pickled report (which contains the dataset), use `report.get_state()`, and in the result replace the dataset by a placeholder / reference. Upload the actual dataset as a separate artifact to benefit from the existing deduplication logic.
2. Making sure we don't do any plots or computations that are not actually needed by the hub, because it does its own plotting.
3. Reducing the number of requests by grouping some of them. For example, we could first round up all the artifacts we need to put and their IDs, send that to the hub, and receive the list of IDs that it doesn't have already + presigned URLs for all of those in 1 request.
4. Parallelizing uploads.

FWICT, @thomass-dev has started on 4., @cakedev0 on 1., and @auguste-probabl 's PR goes in the direction of 3.


I'll also paste here the summary of @cakedev0  about some profiling he did:

> But basically, the put time is an affine function of the report size:
> The intercept time is explained by:
> 
> plotting takes some significant time even for small toy reports
> putting each artifact on the hub (even a tiny one) takes ~1s per artifact (backend latency) 
> (for local project, you have none of those for now, hence put is very fast for small reports)
> The slope is explained mostly by data transfer (to the network or to the disk). This will be dramatically improved by smart serialization.
> The calculation time for metrics and plots is likely negligible compared to that (a few seconds for a report which pickle size is 1GB).

**Comment by cakedev0** [comment]
Note that approaches 3 and 4 are alternatives to solve roughly the same problem. I think 3 is really promising, as backed by some profiling done in the comments of this PR: https://github.com/probabl-ai/skore-hub/pull/1476 (read @auguste-probabl comment and then my comment).

**Comment by auguste-probabl** [comment]
For approach 2, the Hub has implemented custom plots for the ROC curve, PR curve and confusion matrix. However they are only used for binary classification tasks. Still, it's a win in that case.

One potentially big source of plots is the TableReport, in which we compute one plot for every column in the dataset. Contrary to the other plots, the TableReport component has logic implemented to show you the plot corresponding to a column selected with the mouse, so removing the plots means this column selection feature is meaningless. Because of this I'm more reluctant to drop these plots.

Approach 3 is the subject of #2910 (still in development), so I think I'll make a quick PR to drop the ROC/PR/Conf. mat. for binary classification tasks, and then I'll focus on making #2910 ready for review.

**Comment by cakedev0** [comment]
> The slope is explained mostly by data transfer (to the network or to the disk). This will be dramatically improved by smart serialization.

I'd like to edit this a bit, `report.data.analyze()` could also be dominant for big datasets, see: https://github.com/probabl-ai/skore/issues/2939

**Comment by jeromedockes** [comment]
> removing the plots means this column selection feature is meaningless.

there are also some summary statistics computed for each column, so it could still be useful but for some reason they are not shown in the hub

**Comment by auguste-probabl** [comment]
Since #2947 landed I was curious to see the gains.

```py
from skore import evaluate, Project, login
from sklearn.datasets import make_classification
from sklearn.ensemble import HistGradientBoostingClassifier

X, y = make_classification(random_state=0)

# 3-fold CV
report = evaluate(HistGradientBoostingClassifier(random_state=0), X, y, splitter=3)

login()
project = Project("my-workspace/bench-classif", mode="hub")
project.put("test", report)
```
(and similar for regression)

I read the durations directly from the skore progress bar, so the times are only for the `put`.

Results (time measurements in seconds):

|                              | Regression | Binary classification |
|------------------------------|------------|-----------------------|
| 03ed7bd (just before #2947)     | 93         | 50  |
| d46b275 (current `main`)  | 73         | 36 |
| Gain ("x times faster")      | 1.27       | 1.39 |

So the gain is substantial, but nothing to write home about.

I also did a run on `main` with `py-spy` and got this flamegraph:

<img width="1841" height="955" alt="Image" src="https://github.com/user-attachments/assets/ab151ea6-b498-4ac6-952a-b23538848bed" />

The bottleneck now seems to be the TableReport (obtained with `summarize_dataframe`) specifically the plotting.

**Comment by jeromedockes** [comment]
thanks for investigating! (I think the table values are inverted, right? and the link doesn't seem to be to the right pr)

so the next step is clearly spending less time in tablereport: doing it only once, maybe skipping plots, etc

**Comment by auguste-probabl** [comment]
> (I think the table values are inverted, right? and the link doesn't seem to be to the right pr)

Thanks, fixed
